src/java.base/share/classes/sun/security/x509/BasicConstraintsExtension.java

```diff
@@ -119,85 +119,74 @@ public BasicConstraintsExtension(Boolean critical, boolean ca, int len)
-         /*
-          * Activate this check once again after PKIX profiling
-          * is a standard and this check no longer imposes an
-          * interoperability barrier.
-          * if (ca) {
-          *   if (!this.critical) {
-          *   throw new IOException("Criticality cannot be false for CA.");
-          *   }
-          * }
-          */
-     }
+        if (val.data.seeOptional(DerValue.tag_Boolean)) {
+            this.ca = val.data.getBoolean();
+            if (!this.ca) {
+                throw new IOException("Default cA encoded in BasicConstraints");
+            }
+        }
+        if (this.ca) {
+            if (val.data.seeOptional(DerValue.tag_Integer)) {
+                this.pathLen = val.data.getInteger();
+            } else {
+                this.pathLen = Integer.MAX_VALUE;
+            }
+        }
+        val.data.atEnd();
+        /*
+         * Activate this check once again after PKIX profiling
+         * is a standard and this check no longer imposes an
+         * interoperability barrier.
+         * if (ca) {
+         *   if (!this.critical) {
+         *   throw new IOException("Criticality cannot be false for CA.");
+         *   }
+         * }
+         */
+    }
 
-     /**
-      * Return user readable form of extension.
-      */
-     public String toString() {
-         return super.toString() +
-             "BasicConstraints:[\n  CA:" + ca +
-             "\n  PathLen:" +
-             ((pathLen >= 0) ? String.valueOf(pathLen) : " undefined") +
-             "\n]\n";
-     }
+    /**
+     * Return user readable form of extension.
+     */
+    public String toString() {
+        return super.toString() +
+                "BasicConstraints:[\n  CA:" + ca +
+                "\n  PathLen:" +
+                ((pathLen >= 0) ? String.valueOf(pathLen) : " undefined") +
+                "\n]\n";
+    }
 
-     /**
-      * Encode this extension value to the output stream.
-      *
-      * @param out the DerOutputStream to encode the extension to.
-      */
-     public void encode(OutputStream out) throws IOException {
-         DerOutputStream tmp = new DerOutputStream();
-         if (extensionValue == null) {
-             this.extensionId = PKIXExtensions.BasicConstraints_Id;
-             if (ca) {
-                 critical = true;
-             } else {
-                 critical = false;
-             }
-             encodeThis();
-         }
-         super.encode(tmp);
+    /**
+     * Encode this extension value to the output stream.
+     *
+     * @param out the DerOutputStream to encode the extension to.
+     */
+    public void encode(OutputStream out) throws IOException {
+        DerOutputStream tmp = new DerOutputStream();
+        if (extensionValue == null) {
+            this.extensionId = PKIXExtensions.BasicConstraints_Id;
+            if (ca) {
+                critical = true;
+            } else {
+                critical = false;
+            }
```

Comment:
Not relevant to your changes, but it seems strange that the 'critical' value is set solely based on 'ca'. There is a constructor which takes a boolean argument for 'critical'. What is the point of doing that given the "if (ca)" block here?

Reply:
In fact, I don't know how useful this extension is if ca is false. Maybe the non-critical false flag can be comprehended as "I told you I am not a CA, but if you insist to treat me like one, you can pretend you don't understand my words". How about we leave this unchanged?